include/util/json_renderer.hpp

```diff
@@ -4,191 +4,165 @@
 #ifndef JSON_RENDERER_HPP
 #define JSON_RENDERER_HPP
 
-#include "util/cast.hpp"
-#include "util/ieee754.hpp"
 #include "util/string_util.hpp"
 
 #include "osrm/json_container.hpp"
 
+#include <algorithm>
+#include <fmt/format.h>
 #include <iterator>
 #include <ostream>
 #include <string>
 #include <vector>
 
+#include <fmt/compile.h>
+
 namespace osrm
 {
 namespace util
 {
 namespace json
 {
 
-namespace
-{
-constexpr int MAX_FLOAT_STRING_LENGTH = 256;
-}
-
-struct Renderer
+template <typename Out> struct Renderer
 {
-    explicit Renderer(std::ostream &_out) : out(_out) {}
+    explicit Renderer(Out &_out) : out(_out) {}
 
-    void operator()(const String &string) const
+    void operator()(const String &string)
     {
-        out << "\"";
-        out << escape_JSON(string.value);
-        out << "\"";
+        write('"');
+        // here we assume that vast majority of strings don't need to be escaped,
+        // so we check it first and escape only if needed
+        auto size = SizeOfEscapedJSONString(string.value);
+        if (size == string.value.size())
+        {
+            write(string.value);
+        }
+        else
+        {
+            std::string escaped;
+            escaped.reserve(size);
+            EscapeJSONString(string.value, escaped);
+
+            write(escaped);
+        }
+        write('"');
     }
 
-    void operator()(const Number &number) const
+    void operator()(const Number &number)
     {
-        char buffer[MAX_FLOAT_STRING_LENGTH] = {'\0'};
-        ieee754::dtoa_milo(number.value, buffer);
-
-        // Trucate to 10 decimal places
-        int pos = 0;
-        int decimalpos = 0;
-        while (decimalpos == 0 && pos < MAX_FLOAT_STRING_LENGTH && buffer[pos] != 0)
-        {
-            if (buffer[pos] == '.')
-            {
-                decimalpos = pos;
-                break;
-            }
-            ++pos;
-        }
-        while (pos < MAX_FLOAT_STRING_LENGTH && buffer[pos] != 0)
+        // `fmt::memory_buffer` stores first 500 bytes in the object itself(i.e. on stack in this
+        // case) and then grows using heap if needed
+        fmt::memory_buffer buffer;
+        fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{}"), number.value);
+
+        // Truncate to 10 decimal places
```

Comment:
Tbh I do not understand why do we truncate it - I couldn't find nothing about it in JSON spec. Just to save some bytes in corner cases?

Reply:
I'd assume it's because none of the API responses can attribute anything meaningful to more decimal places, so it saves response payload space + improves readability.